src/server/pfs/s3/bucket.go

```diff
@@ -0,0 +1,396 @@
+			}
+			result.Contents = append(result.Contents, contents)
+		} else {
+			result.CommonPrefixes = append(result.CommonPrefixes, newCommonPrefixes(fileInfo.File.Path))
+		}
+		return nil
+	})
+
+	if err != nil {
+		newInternalError(r, err).write(w)
+		return
+	}
+
+	writeXML(w, http.StatusOK, result)
+}
+
+func (h bucketHandler) listBranch(w http.ResponseWriter, r *http.Request, result *ListBucketResult, branch string, filePrefix string) {
+	// ensure that we can globify the prefix string
+	if !isGlobless(filePrefix) {
+		newGlobbyPrefixError(r).write(w)
+		return
+	}
+
+	fileInfos, err := h.pc.GlobFile(result.Name, branch, fmt.Sprintf("%s*", filePrefix))
+	if err != nil {
+		newInternalError(r, err).write(w)
+		return
+	}
+
+	for _, fileInfo := range fileInfos {
+		fileInfo = updateFileInfo(branch, result.Marker, fileInfo)
+		if fileInfo == nil {
+			continue
+		}
+		if result.isFull() {
+			result.IsTruncated = true
+			break
+		}
+		if fileInfo.FileType == pfs.FileType_FILE {
+			contents, err := newContents(fileInfo)
+			if err != nil {
+				newInternalError(r, err).write(w)
+				return
+			}
+			result.Contents = append(result.Contents, contents)
+		} else {
+			result.CommonPrefixes = append(result.CommonPrefixes, newCommonPrefixes(fileInfo.File.Path))
+		}
+	}
+
+	writeXML(w, http.StatusOK, result)
+}
+
+func (h bucketHandler) put(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	repo := vars["repo"]
+
+	err := h.pc.CreateRepo(repo)
+	if err != nil {
+		if strings.Contains(err.Error(), "as it already exists") {
+			newBucketAlreadyExistsError(r).write(w)
+		} else {
+			newInternalError(r, err).write(w)
+		}
+
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+}
+
+func (h bucketHandler) del(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	repo := vars["repo"]
+
+	err := h.pc.DeleteRepo(repo, false)
+
+	if err != nil {
+		newNotFoundError(r, err).write(w)
+		return
+	}
+
+	w.WriteHeader(http.StatusNoContent)
+}
+
+// updateFileInfo takes in a `FileInfo`, and updates it to be used in s3
+// object listings:
+// 1) if nil is returned, the `FileInfo` should not be included in the list
+// 2) the path is updated to be prefixed by the branch name
+func updateFileInfo(branch, marker string, fileInfo *pfs.FileInfo) *pfs.FileInfo {
+	if fileInfo.FileType != pfs.FileType_FILE && fileInfo.FileType != pfs.FileType_DIR {
+		// skip anything that isn't a file or dir
+		return nil
+	}
+	if fileInfo.FileType == pfs.FileType_DIR && fileInfo.File.Path == "/" {
+		// skip the root directory
+		return nil
+	}
+
+	// update the path to match s3
+	fileInfo.File.Path = fmt.Sprintf("%s%s", branch, fileInfo.File.Path)
+
+	if fileInfo.File.Path <= marker {
+		// skip file paths below the marker
+		return nil
+	}
+
+	return fileInfo
+}
+
+// isGlobless checks whether any a string contains any characters used in
+// glob file path matching
+func isGlobless(s string) bool {
+	return !strings.ContainsAny(s, globSpecialCharacters)
+}
```

Comment:
We have a function `IsGlob` at `src/server/pkg/hashtree/db.go:1474`. Probably better to reuse that here.

Reply:
Looks like the glob package we pull in actually has a way of escaping glob characters, so I'll use that to avoid another divergence from S3